File: rust/src/pol.rs
use std::io::{Write, stdin};

/*
 * Llegir polinomi de fitxer
 */
pub fn plleg() -> (i32, Option<Vec<f64>>) {
    let mut line = String::new();
    if stdin().read_line(&mut line).is_err() {
        eprintln!("plleg(): error llegint grau !!");
        return (-1, None);
    }

    let n = match line.trim().parse::<i32>() {
        Ok(val) => val,
        Err(_) => {
            eprintln!("plleg(): error llegint grau !!");
            return (-1, None);
        }
    };

    let mut vec = vec![0.0; (n + 1) as usize];

    for i in 0..=n {
        line.clear();
        if stdin().read_line(&mut line).is_err() {
            eprintln!("plleg(): error llegint coeficient !!");
            return (-3, None);
        }

        vec[i as usize] = match line.trim().parse::<f64>() {
            Ok(val) => val,
            Err(_) => {
                eprintln!("plleg(): error llegint coeficient !!");
                return (-3, None);
            }
        };
    }

    (n, Some(vec))
}

/*
 * Escriure polinomi a fitxer.
 */
pub fn pescr<W: Write>(n: i32, p: &Vec<f64>, fp: &mut W) {
    writeln!(fp, "{}", n).ok();
    for coef in p.iter().take((n + 1) as usize) {
        writeln!(fp, "{}", coef).ok();
    }
}

/*
 * Combinació lineal de polinomis.
 *
 * La funció que crida és responsable que r[] tingui lloc per
 * 1+max(m,n) coeficients.
 */
pub fn pcl(m: i32, p: &Vec<f64>, n: i32, q: &Vec<f64>, alf: f64, bet: f64, r: &mut [f64]) {
    let m_usize = (m + 1) as usize;
    let n_usize = (n + 1) as usize;

    for (i, val) in p.iter().take(m_usize).enumerate() {
        r[i] = alf * val;
    }
    for (i, val) in q.iter().take(n_usize).enumerate() {
        r[i] += bet * val;
    }
}

/*
 * Producte de polinomis.
 *
 * La funció que crida és responsable que r[] tingui lloc per
 * 1+m+n coeficients.
 */
pub fn pmul(m: i32, p: &Vec<f64>, n: i32, q: &Vec<f64>) -> Vec<f64> {
    let m_usize = (m + 1) as usize;
    let n_usize = (n + 1) as usize;
    let mut r = vec![0.0; m_usize + n_usize];

    for (i, p_val) in p.iter().take(m_usize).enumerate() {
        for (j, q_val) in q.iter().take(n_usize).enumerate() {
            r[i + j] += p_val * q_val;
        }
    }
    r
}

/*
 * Derivada d'un polinomi
 *
 * La funció que crida és responsable que dp[] tingui lloc
 * per a n coeficients.
 */
pub fn pder(n: i32, p: &Vec<f64>) -> Vec<f64> {
    let mut dp = vec![0.0; n as usize];
    for (i, val) in p.iter().skip(1).take(n as usize).enumerate() {
        dp[i] = ((i + 1) as f64) * val
    }
    dp
}

/*
 * Avaluar un polinomi.
 */
pub fn paval(n: i32, p: &Vec<f64>, x: f64) -> f64 {
    p.iter()
        .take((n + 1) as usize)
        .enumerate()
        .map(|(i, &coef)| coef * x.powi(i as i32))
        .sum()
}

File: pol.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Llegir polinomi de fitxer
 */
#define ERROR(msg,ret) \
    fprintf(stderr,"plleg(): error " msg " !!\n");\
    iret=ret;\
    goto allib\

int plleg (FILE *fp, int *n, double **p) {
    int iret, i;
    iret=0;
    (*p)=NULL;
    if (fscanf(fp,"%d",n)!=1) { ERROR("llegint grau",-1); }
    (*p)=malloc(( (*n)+1 )*sizeof(double));
    if ((*p)==NULL) { ERROR("reservant memòria",-2); }
    for (i=0; i<=(*n); i++)
        if (fscanf(fp,"%lf",&(*p)[i])!=1) { ERROR("llegint coeficient",-3); }
    iret=0; goto fi;
    allib:
        if (*p!=NULL) free(*p);
    fi:
        return iret;
}

#undef ERROR

/*
 * Escriure polinomi a fitxer.
 */
void pescr (int n, double *p, FILE *fp) {
    int i;
    fprintf(fp,"%d\n",n);
    for (i=0; i<=n; i++) fprintf(fp,"%.16g\n",p[i]);
}

/*
 * Combinació lineal de polinomis.
 *
 * La funció que crida és responsable que r[] tingui lloc per
 * 1+max(m,n) coeficients.
 */
void pcl (int m, double *p, int n, double *q, double alf, double bet, double r[]) {
    int i, max = n;
    if (m > n) max = m;
    for (i=0; i<=max; i++) r[i] = 0;
    for (i=0; i<=m; i++) r[i] += alf*p[i];
    for (i=0; i<=n; i++) r[i] += bet*q[i];
}

/*
 * Producte de polinomis.
 *
 * La funció que crida és responsable que r[] tingui lloc per
 * 1+m+n coeficients.
 */
void pmul (int m, double *p, int n, double *q, double *r) {
    int i, j;
    for (i=0; i<=n+m; i++) r[i] = 0;
    for (i=0;i<=m;i++)
        for (j=0;j<=n;j++)
            r[i+j] += p[i]*q[j];
}

/*
 * Derivada d'un polinomi
 *
 * La funció que crida és responsable que dp[] tingui lloc
 * per a n coeficients.
 */
void pder (int n, double *p, double *dp) {
    dp[0] = 0;
    for (int i = 1; i <= n; i++)
        dp[i-1] = i*p[i];
}

/*
 * Avaluar un polinomi.
 */
double paval (int n, double *p, double x) {
    double y;
    y = 0;
    for (int i = n; 0 <= i; i--) {
        y = p[i] + y*x;
    }
    return y;
}
